osd: initial minimal efforts to clean up PG interface

liewegas opened the pull request:
@branch-predictor the biggest change here is the clean up of the forced recovery interface

joscollin commented:
@liewegas 

```
/home/jenkins-build/build/workspace/ceph-pull-requests/src/osd/PG.h:2443:15: error: ‘static bool PG::_has_removal_flag(ObjectStore*, spg_t)’ is protected
   static bool _has_removal_flag(ObjectStore *store, spg_t pgid);
               ^
/home/jenkins-build/build/workspace/ceph-pull-requests/src/tools/ceph_objectstore_tool.cc:379:62: error: within this context
        (it->is_pg(&pgid) && PG::_has_removal_flag(store, pgid))) {
                                                              ^
```

branch-predictor commented:
@liewegas I don't like producing a new log line for each of pg involved (and right now we even don't know which pgs are involved), but otherwise changes to forced recovery seem OK.

liewegas commented:
It's at debug level 10... should only be seen by developers?

branch-predictor commented:
@liewegas true, but if something breaks and you debug production, you'd prefer to increase log level for a few minutes over attaching gdb to running osd, risking I/O stalls that could be hurting cluster users.

liewegas commented:
I switched it to 20.. is that better?  I think it's helpful to print when we actually switch the state.  If necessary we can add back the pgid list in the caller, but i'm not sure it's needed?

branch-predictor commented:
10 or 20 doesn't make much of difference as long as it's there. In case the pgs won't get forced flags, you may need to find out why, whether that was wrongly issued command that missed some pgs (for example, a bug in some script that calls forced recovery) or bad timing that caused pgs in question to change state to one that's invalid for these commands *after* issuing them.

liewegas commented:
The last commit adds a single debug line with a summary.. does that help?

branch-predictor commented:
Yeah, looks reasonably except in case when the "did" set is empty.

tchaikov commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -11018,22 +11018,22 @@ bool PrimaryLogPG::start_recovery_ops(
   if (unfound) {
     dout(10) << " still have " << unfound << " unfound" << dendl;
-    return work_in_progress;
+    return !work_in_progress;
```
if `work_in_progress` is true, we will return at https://github.com/ceph/ceph/pull/17708/commits/2b432c1199652887b0a662b4f0344e208304f165#diff-fb41013d27e932534adb50eb3de2aaa5R11007, so `!work_in_progress` always evaluates to `true` here.

tchaikov replied:
@liewegas not addressed

liewegas replied:
it's fixed in ea32e82e6e4a52ef4c6728d36dfc747b053c5edb, not sure why github is still showing this

tchaikov commented on src/osd/PG.cc:
```diff
@@ -5817,6 +5817,7 @@ void PG::handle_loaded(RecoveryCtx *rctx)
   dout(10) << "handle_loaded" << dendl;
   Load evt;
   recovery_state.handle_event(evt, rctx);
+  write_if_dirty(*rctx->transaction);
```
i am not sure if this works. in `OSD::load_pgs()`, a nullptr is passed in as the transaction of rctx. like
```c++
    PG::RecoveryCtx rctx(0, 0, 0, 0, 0, 0);
    pg->handle_loaded(&rctx);
```
can we dereference it here w/o checking it beforehand?

liewegas replied:
fixed this one already

liewegas commented:
http://pulpito.ceph.com/sage-2017-10-03_21:51:15-rados-wip-pg-distro-basic-smithi/

liewegas commented on src/osd/OSD.cc:
```diff
@@ -4396,19 +4396,11 @@ void OSD::maybe_update_heartbeat_peers()
-	if (osdmap->is_up(*p))
-	  _add_heartbeat_peer(*p);
-      pg->heartbeat_peer_lock.Unlock();
+      pg->with_heartbeat_peers([&](int peer) {
```
the lamba takes peer, but it's called on all peers.

tchaikov replied:
the function takes a single peer not a set of peers, so it should be `with_heartbeat_peer()`. but we can leave it for another change if you like.

tchaikov commented on src/osd/PG.cc:
```diff
@@ -8204,6 +8204,23 @@ void PG::dump_pgstate_history(Formatter *f)
+    for (auto l : missing_loc.get_locations(i.first)) {
+      f->dump_object("shard", l);
+    }
+    f->close_section();
```
@liewegas not addressed

tchaikov commented on src/osd/PG.h:
```diff
@@ -399,15 +405,65 @@ class PG : public DoutPrefixProvider {
+  }
+
+
+protected:
```
@liewegas not addressed

liewegas commented:
@tchaikov I think I've addressed everything except for the cosmetic squash.

http://pulpito.ceph.com/sage-2017-10-05_05:59:32-rados-wip-sage-testing-2017-10-04-2208-distro-basic-smithi/

My only remaining concern here is that I've removed the past_intervals regeneration code from the OSD and the ceph-objectstore-tool rm-past-intervals code that used to test it.  I don't think it's needed... but maybe I'm missing something?  On the upside if it is needed it has to be rewritten anyway to respect the PG interface, so there's no effort saved by doing it now.  @jdurgin @dzafman thoughts?

liewegas commented:
@dzafman @jdurgin Okay, got to the bottom of the import vs osdmap epoch thing.

Previously,
- import blindly set pg epoch to osd's latest
- import cleared same_interval_since, which told the osd to rebuild history and past intervals
- osd would rebuild on startup

Now,
- cot will fail if the pg file's epoch is < osd's oldest_map.  unless --force is passed, but that means the PastIntervals won't be 100% accurate.  (This is no different than before; it's just now explicit in cot instead of somethign you might notice by digging through the OSD log.  In practice nobody is likely to be importing PGs from before the cluster was last clean.)
- cot will walk the pg through osdmaps up to current_epoch, updating history (same_*_since) and PastIntervals

The net result is that the imported PG will already be up to date when the OSD starts, vs previously where it would have to do a rebuild on startup and there would be some hidden activity slowing osd startup.

dzafman commented:
@liewegas The last commits particularly the changes to ceph_objectstore_tool.cc look good to me.

liewegas commented:
http://pulpito.ceph.com/sage-2017-10-08_01:05:08-rados-wip-sage-testing-2017-10-07-1112-distro-basic-smithi/
http://pulpito.ceph.com/sage-2017-10-08_19:01:45-rados-wip-sage-testing-2017-10-07-1112-distro-basic-smithi/

smithfarm commented:
@liewegas This PR is pretty big, but its absence in luminous is causing quite some conflicts - see http://tracker.ceph.com/issues/22069 . . . @dzafman suggests backporting this PR to luminous - do you agree?

smithfarm commented:
(Optimistically opened http://tracker.ceph.com/issues/22423 for the backport.)

liewegas commented:
It makes me very nervous to backport this.  Between luminous and this PR a ton of compatibility code for <luminous was ripped out, so I'm worried that there is some dependency (although I guess that will turn on during a backport).  

FWIW I was planning on backporting the recovery reservation fixes and it was too painful... maybe this would help, but I'm also okay with not backporting it at all.

I guess give it a go and see how it goes?

liewegas commented:
For example, https://github.com/ceph/ceph/pull/17708/commits/72d50e463f4931ecddc5d859231e55a2b1e768df can't be backported.

liewegas commented:
I think the past intervals stuff shoudl backport okay.  Just pay attention when you're working through the patches for anything that's needed for kraken and jewel compat!
